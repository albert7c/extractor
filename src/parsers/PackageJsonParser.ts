import { Dependency, DependencyManagementFile } from "../../api/deps";
import {JsonParser} from "./Parser";

interface ID {
    organization: string;
    module: string;
}

function parseName(module: string): ID {
    let organization = "_";
    if (module.charAt(0) === "@") {
        const index = module.indexOf("/");
        organization = module.substr(1,  index - 1);
        module = module.substr(index + 1);
    }
    return { organization, module };
}

function extract(dependencyHash: any, scope: string): Dependency[] {
    return Object.keys(dependencyHash)
        .map((dependency) => {
            const { organization, module } = parseName(dependency);
            const versionConstraint = dependencyHash[dependency];

            return {
                organization,
                module,
                versionConstraint,
                scopes: [ scope ],
            };
        });
}

export default class PackageJsonParser extends JsonParser {
    public pathMatch(path: string): boolean {
        return path.endsWith("package.json") && path.indexOf("node_modules") === -1;
    }

    public parseJson({
        name,
        version,
        dependencies,
        devDependencies,
        peerDependencies,
        bundledDependencies,
        optionalDependencies,
    }): DependencyManagementFile {
        const { organization, module } = parseName(name);

        let allDependencies = extract((dependencies || {}), "");
        allDependencies = allDependencies.concat(extract((devDependencies || {}), "dev"));
        allDependencies = allDependencies.concat(extract((peerDependencies || {}), "peer"));
        // deps = deps.concat(extract((bundledDependencies || {}), "bundled"));
        allDependencies = allDependencies.concat(extract((optionalDependencies || {}), "optional"));

        return {
            language: "node",
            system: "npm",
            organization, module, version,
            dependencies: allDependencies,
        };
    }

}
