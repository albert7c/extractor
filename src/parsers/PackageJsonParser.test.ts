import {readFile} from "fs";
import {promisify} from "util";
import PackageJsonParser from "./PackageJsonParser";

const readFileAsync = promisify(readFile);

describe("PackageJsonParser", () => {
    test("fullParse", async () => {
        const jsonPath = require.resolve("./testdata/package.json");
        const buffer = await readFileAsync(jsonPath);
        const content = buffer.toString();

        const parser = new PackageJsonParser();

        const actual = parser.parse(jsonPath, content);

        expect(actual).toMatchSnapshot();
        expect(JSON.stringify(actual, null, 2)).toMatchSnapshot();
    });
});
