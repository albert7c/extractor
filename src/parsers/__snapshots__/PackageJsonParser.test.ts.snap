// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PackageJsonParser fullParse 1`] = `
Object {
  "dependencies": Array [
    Object {
      "module": "runtimeDependency",
      "organization": "scoped",
      "scopes": Array [
        "",
      ],
      "versionConstraint": "1.0.0",
    },
    Object {
      "module": "devDependency",
      "organization": "_",
      "scopes": Array [
        "dev",
      ],
      "versionConstraint": "1.0.0",
    },
    Object {
      "module": "peerDependency",
      "organization": "scoped",
      "scopes": Array [
        "peer",
      ],
      "versionConstraint": "1.0.0",
    },
    Object {
      "module": "optionalDependency",
      "organization": "_",
      "scopes": Array [
        "optional",
      ],
      "versionConstraint": "1.0.0",
    },
  ],
  "language": "node",
  "module": "module",
  "organization": "organization",
  "system": "npm",
  "version": "9.9.9",
}
`;

exports[`PackageJsonParser fullParse 2`] = `
"{
  \\"language\\": \\"node\\",
  \\"system\\": \\"npm\\",
  \\"organization\\": \\"organization\\",
  \\"module\\": \\"module\\",
  \\"version\\": \\"9.9.9\\",
  \\"dependencies\\": [
    {
      \\"organization\\": \\"scoped\\",
      \\"module\\": \\"runtimeDependency\\",
      \\"versionConstraint\\": \\"1.0.0\\",
      \\"scopes\\": [
        \\"\\"
      ]
    },
    {
      \\"organization\\": \\"_\\",
      \\"module\\": \\"devDependency\\",
      \\"versionConstraint\\": \\"1.0.0\\",
      \\"scopes\\": [
        \\"dev\\"
      ]
    },
    {
      \\"organization\\": \\"scoped\\",
      \\"module\\": \\"peerDependency\\",
      \\"versionConstraint\\": \\"1.0.0\\",
      \\"scopes\\": [
        \\"peer\\"
      ]
    },
    {
      \\"organization\\": \\"_\\",
      \\"module\\": \\"optionalDependency\\",
      \\"versionConstraint\\": \\"1.0.0\\",
      \\"scopes\\": [
        \\"optional\\"
      ]
    }
  ]
}"
`;
